pkg/engine2/cli.go

```diff
@@ -256,22 +264,107 @@ func setupProfiling() func() {
 	return func() {}
 }
 
-func (em *EngineMain) RunEngine(cmd *cobra.Command, args []string) error {
+func (em *EngineMain) Run(context *EngineContext) (int, []engine_errs.EngineError) {
+	returnCode := 0
+	var engErrs []engine_errs.EngineError
+
+	zap.S().Info("Running engine")
+	err := em.Engine.Run(context)
+	if err != nil {
+		returnCode = 1
+		if ee, ok := err.(engine_errs.EngineError); ok {
+			engErrs = append(engErrs, ee)
+		} else {
+			engErrs = append(engErrs, engine_errs.InternalError{Err: engine_errs.ErrorsToTree(err)})
+		}
+	}
+
+	if len(context.Solutions) > 0 {
+		writeDebugGraphs(context.Solutions[0])
+		for _, d := range context.Solutions[0].GetDecisions().GetRecords() {
+			d, ok := d.(solution_context.MaybeErroDecision)
+			if !ok {
+				continue
+			}
+			ee := d.AsEngineError()
+			if ee == nil {
+				continue
+			}
+			engErrs = append(engErrs, ee)
+			if returnCode != 1 {
+				returnCode = 2
+			}
+		}
+	}
+
+	return returnCode, engErrs
+}
+
+func writeEngineErrsJson(errs []engine_errs.EngineError, out io.Writer) error {
+	enc := json.NewEncoder(out)
+	enc.SetIndent("", "  ")
+	// NOTE: since this isn't used in a web context (it's a CLI), we can disable escaping.
+	enc.SetEscapeHTML(false)
+
+	outErrs := make([]map[string]any, len(errs))
+	for i, e := range errs {
+		outErrs[i] = e.ToJSONMap()
+		outErrs[i]["error_code"] = e.ErrorCode()
+		wrapped := errors.Unwrap(e)
+		if wrapped != nil {
+			outErrs[i]["error"] = engine_errs.ErrorsToTree(wrapped)
+		}
+	}
+	return enc.Encode(outErrs)
+}
+
+func (em *EngineMain) RunEngine(cmd *cobra.Command, args []string) (exitCode int) {
+	var engErrs []engine_errs.EngineError
+	internalError := func(err error) {
+		engErrs = append(engErrs, engine_errs.InternalError{Err: err})
+		exitCode = 1
+	}
+
+	defer func() { // defer functions execute in FILO order, so this executes after the 'recover'.
+		err := writeEngineErrsJson(engErrs, os.Stdout)
+		if err != nil {
+			zap.S().Errorf("failed to output errors to stdout: %v", err)
+		}
+	}()
+	defer func() {
+		r := recover()
+		if r == nil {
+			return
+		}
+		zap.S().Errorf("panic: %v", r)
+		switch r := r.(type) {
+		case engine_errs.EngineError:
+			engErrs = append(engErrs, r)
```

Comment:
should an engine error ever cause a panic?

Reply:
It _shouldn't_, but in case it does I wouldn't want to wrap it in an internal error. I figured the check was worth the extra 2 line, even if it doesn't happen right now.